pwndbg/aglib/heap/mallocng.py

```diff
@@ -527,11 +581,465 @@ def slot_size(self):
             # The meta is corrupted.
             return -1
 
+    @staticmethod
+    def sizeof():
+        return 0x28
+
 
 class MetaArea:
+    """
+    Slabs that contain metas, linked in a singly-linked list.
+
+    https://elixir.bootlin.com/musl/v1.2.5/source/src/malloc/mallocng/meta.h#L34
+    struct meta_area {
+      uint64_t check;
+      struct meta_area *next;
+      int nslots;
+      struct meta slots[];
+    };
+    """
+
     def __init__(self, addr: int) -> None:
-        self.addr = addr
+        self.addr: int = addr
+
+        self.check: int = 0
+        self.meta_area: int = 0
+        self.nslots: int = 0
+        # Alignment offsets it by 0x4
+        self.slots: int = self.addr + 0x18
+
+        self.load()
+
+    def load(self):
+        ptrsize = pwndbg.aglib.arch.ptrsize
+        uint64size = pwndbg.aglib.typeinfo.uint64.sizeof
+        endian = pwndbg.aglib.arch.endian
+
+        data: bytearray = memory.read(self.addr, 0x14)
+
+        cur_offset = 0
+        self.check = int.from_bytes(
+            data[cur_offset : (cur_offset + uint64size)], endian, signed=False
+        )
+        cur_offset += uint64size
+        self.next = pwndbg.aglib.arch.unpack(data[cur_offset : (cur_offset + ptrsize)])
+        cur_offset += ptrsize
+        self.nslots = int.from_bytes(
+            data[cur_offset : (cur_offset + int_size())], endian, signed=True
+        )
+        cur_offset += int_size()
+
+        assert cur_offset == 0x14
+
+    def at_index(self, idx: int) -> int:
+        """
+        Returns the address of the meta object located
+        at index idx.
+        """
+        return self.addr + 0x18 + idx * Meta.sizeof()
```

Comment:
That's at least 2nd time we use 0x18 - shall we move it to a constant?

Reply:
I don't know what these hardcoded values represent, but are they valid for architectures that use different width pointers?